crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,411 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::bigtable::proto::bigtable::v2::bigtable_client::BigtableClient as BigtableInternalClient;
+use crate::bigtable::proto::bigtable::v2::mutate_rows_request::Entry;
+use crate::bigtable::proto::bigtable::v2::mutation::SetCell;
+use crate::bigtable::proto::bigtable::v2::read_rows_response::cell_chunk::RowStatus;
+use crate::bigtable::proto::bigtable::v2::{
+    mutation, MutateRowsRequest, MutateRowsResponse, Mutation, ReadRowsRequest, RowSet,
+};
+use crate::{Checkpoint, KeyValueStoreReader, KeyValueStoreWriter, TransactionData};
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use gcp_auth::{Token, TokenProvider};
+use http::{HeaderValue, Request, Response};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+use std::task::{Context, Poll};
+use std::time::Duration;
+use sui_types::base_types::{SequenceNumber, TransactionDigest};
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::object::Object;
+use sui_types::storage::ObjectKey;
+use tonic::body::BoxBody;
+use tonic::codegen::Service;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Streaming;
+use tracing::error;
+
+const OBJECTS_TABLE: &str = "objects";
+const CHECKPOINTS_TABLE: &str = "checkpoints";
+const TRANSACTIONS_TABLE: &str = "transactions";
+
+const COLUMN_FAMILY_NAME: &str = "sui";
+const DEFAULT_COLUMN_QUALIFIER: &str = "";
+const CHECKPOINT_SUMMARY_COLUMN_QUALIFIER: &str = "s";
+const CHECKPOINT_CONTENTS_COLUMN_QUALIFIER: &str = "c";
+const TRANSACTION_COLUMN_QUALIFIER: &str = "tx";
+const EFFECTS_COLUMN_QUALIFIER: &str = "ef";
+const EVENTS_COLUMN_QUALIFIER: &str = "ev";
+const TIMESTAMP_COLUMN_QUALIFIER: &str = "ts";
+const CHECKPOINT_NUMBER_COLUMN_QUALIFIER: &str = "cn";
+
+type Bytes = Vec<u8>;
+
+#[derive(Clone)]
+struct AuthChannel {
+    channel: Channel,
+    policy: String,
+    token_provider: Option<Arc<dyn TokenProvider>>,
+    token: Arc<RwLock<Option<Arc<Token>>>>,
+}
+
+#[derive(Clone)]
+pub struct BigTableClient {
+    table_prefix: String,
+    client: BigtableInternalClient<AuthChannel>,
+}
+
+#[async_trait]
+impl KeyValueStoreWriter for BigTableClient {
+    async fn save_objects(&mut self, objects: &[&Object]) -> Result<()> {
+        let mut items = Vec::with_capacity(objects.len());
+        for object in objects {
+            let object_key = ObjectKey(object.id(), object.version());
+            items.push((
+                Self::raw_object_key(&object_key)?,
+                vec![(DEFAULT_COLUMN_QUALIFIER, bcs::to_bytes(object)?)],
```

Comment:
Do we want to be explicit and name the column "object", idk how default vs naming it has an impact on things

Reply:
responded above. Empty qualifier consumes the least space per row and makes sense if table consists of 1 column

Reply:
excellent! thanks for enlightening me on how bigtable works!